### ISSUE TITLE
Parsing parts do not work properly (see description and referenced issues)

### ISSUE THREAD

**Issue by DesarrolloLanalden** [barbushin/php-imap#278]
Hi,

We have some emails that have eml attachments, but when they are parsed using the library the appears as a part of the body instead of an attachment.

The original message is attached,
[originalEmail.txt](https://github.com/barbushin/php-imap/files/2803601/originalEmail.txt)

Any thoughts?

**Comment by Sebi94nbg** [comment]
This issue does still exist with `php-imap` v3.0.11.

The mail structure looks like this in the case of the provided example email:
```
{
	"type": 1,
	"encoding": 0,
	"ifsubtype": 1,
	"subtype": "MIXED",
	"ifdescription": 0,
	"ifid": 0,
	"ifdisposition": 0,
	"ifdparameters": 0,
	"ifparameters": 1,
	"parameters": [{
		"attribute": "boundary",
		"value": "----=_NextPart_000_0007_01D4B71A.CFE48640"
	}],
	"parts": [{
		"type": 1,
		"encoding": 0,
		"ifsubtype": 1,
		"subtype": "ALTERNATIVE",
		"ifdescription": 0,
		"ifid": 0,
		"ifdisposition": 0,
		"ifdparameters": 0,
		"ifparameters": 1,
		"parameters": [{
			"attribute": "boundary",
			"value": "----=_NextPart_001_0008_01D4B71A.CFE48640"
		}],
		"parts": [{
			"type": 0,
			"encoding": 0,
			"ifsubtype": 1,
			"subtype": "PLAIN",
			"ifdescription": 0,
			"ifid": 0,
			"lines": 2,
			"bytes": 39,
			"ifdisposition": 0,
			"ifdparameters": 0,
			"ifparameters": 1,
			"parameters": [{
				"attribute": "charset",
				"value": "us-ascii"
			}]
		},
		{
			"type": 0,
			"encoding": 4,
			"ifsubtype": 1,
			"subtype": "HTML",
			"ifdescription": 0,
			"ifid": 0,
			"lines": 55,
			"bytes": 1895,
			"ifdisposition": 0,
			"ifdparameters": 0,
			"ifparameters": 1,
			"parameters": [{
				"attribute": "charset",
				"value": "us-ascii"
			}]
		}]
	},
	{
		"type": 2,
		"encoding": 0,
		"ifsubtype": 1,
		"subtype": "RFC822",
		"ifdescription": 0,
		"ifid": 0,
		"lines": 28,
		"bytes": 1376,
		"ifdisposition": 1,
		"disposition": "attachment",
		"ifdparameters": 0,
		"ifparameters": 0,
		"parameters": {
			
		},
		"parts": [{
			"type": 0,
			"encoding": 4,
			"ifsubtype": 1,
			"subtype": "PLAIN",
			"ifdescription": 0,
			"ifid": 0,
			"lines": 2,
			"bytes": 142,
			"ifdisposition": 0,
			"ifdparameters": 0,
			"ifparameters": 1,
			"parameters": [{
				"attribute": "charset",
				"value": "utf-8"
			}]
		}]
	}]
}
```

The lines [768-895](https://github.com/barbushin/php-imap/blob/master/src/PhpImap/Mailbox.php#L768-L895) must be improved / fixed in order to fix this issue. These lines are about the functions `getMail()` and `initMailPart()`.

We should base our code on this one and optimize it accordingly: https://stackoverflow.com/questions/11285430/imap-get-attached-file

Another example code: https://www.electrictoolbox.com/extract-attachments-email-php-imap/

The question is: When is a part an attachment? Always, if `disposition` is `attachment`? Or always, when an `attribute` called `filename` exists? I would say when both is given. Any other opinions?

A list of available subtypes can be found at Wikipedia: https://en.wikipedia.org/wiki/MIME#Multipart_subtypes

The german version of this site shows the types better: https://de.wikipedia.org/wiki/Multipurpose_Internet_Mail_Extensions#MIME_Part_2_%E2%80%93_Media_Types

Information regarding the object, body types and transfer encodings of `imap_fetchstructure()` can be found here: https://www.php.net/manual/en/function.imap-fetchstructure.php

Any pull requests to improve this are welcome.

I'll working on a fix as well.

**Comment by EndikaRamos** [comment]
Hello,

My litte grain of sand :

Acoording to https://tools.ietf.org/html/rfc2183

> 2.3  The Filename Parameter
> 
>    The sender may want to suggest a filename to be used if the entity is
>    detached and stored in a separate file. 

So, I think tha according to rfc2183, the filename is not a "MUST". Does it means that a filename is not needed por the part to be a parsed as an attachment?

>    In addition to allowing the sender to specify the presentational
>    disposition of a message component, it is desirable to allow her to
>    indicate a default archival disposition; a filename. The optional
>    "filename" parameter provides for this. 

About, the eml file, I've found, a similar probano delem in the following library:
https://github.com/roundcube/roundcubemail/issues/3466

It seems, that is the attachment subtype is RFC822, and the disposition is attachment, it usually contains  an "eml file".

References:
https://stackoverflow.com/questions/21936983/imap-parsing-original-headers-from-a-bounced-mail
https://www.w3.org/Protocols/rfc1341/rfc1341.html
https://stackoverflow.com/questions/38874376/javamail-also-extract-attachments-of-encapsulated-message-content-type-message

This also, leaves another question, if we manage to download the eml file. Should it be downloaded as an eml (with headers and other attached attachments)? Or... should it be downloaded as html/plain and any other attacment contained inside de eml also dowloaded separately?

**Comment by Sebi94nbg** [comment]
Yeah, I already thought, that the file name is not required. The EML file for example does also not have a filename, even when a mail client like Outlook shows a filename. :(

I would leave attached EML files as attachment, but we may add a boolean flag to let the user decide, if he wants the attached EML file as it is (with headers) or in decoded format (HTML/PLAIN).

Thanks for your research and the input, @EndikaRamos!

**Comment by Sebi94nbg** [comment]
I think, I've found a solution. I just need to implement my code into `php-imap` to fill the properties and to test this change then. Hopefully, this is done by tomorrow end of day. :)

**Comment by Sebi94nbg** [comment]
Currently trying to detect attachments based on this section information: https://www.php.net/manual/en/function.imap-fetchbody.php#89002

**Comment by Sebi94nbg** [comment]
Fixing this issue should also solve this issue: #198

**Comment by Sebi94nbg** [comment]
I've created a new branch called `optimaze-part-parsing` with my latest code changes. (see reference above)

I haven't committed it into develop yet, because it's not working as expected.

Parsing the parts is in the most cases returning the correct part as "is an attachment", but sometimes it also says this to parts, which are not an attachment. The class `DataPartInfo` also needs improvements - it is sometimes parsing attachments as body, sometimes it's not parsing some inline attachments at all and sometimes it's not even parsing the body.

Maybe someone else has an idea, where the worm is hidden in this code. Maybe @nickl- as he added this class?

@Everyone: Feel free to checkout this new branch, to change whatever you need to change, to test and commit your changes: `composer require php-imap/php-imap:dev-optimaze-part-parsing`

**Comment by EndikaRamos** [comment]
Although, it is not the ideal solution for this problem, but I've been able to make it works with eml files,

Using the actual branch (develop), I've been able to:

Download de the eml file as ".eml" attached file
Download each file "inside" fhe eml as a different attachment
"Mark" (as a parameter called "emlOrigin") if the attachment comes from an eml files
Example, of processed EML "attachment":

10431_5845944131528768489_5845944131528768489.rfc822.eml
10431_9025868031894177481_9025868031894177481.plain
@Sebi94nbg , what do you think? Any problems than colud arise with this solution?

To do this, the code to download the attchment has been split into another function (downloadAttachment). If it is ok, could someone help me with the comments? English is not my primary language.

https://github.com/barbushin/php-imap/pull/341

**Comment by Sebi94nbg** [comment]
Thanks for the PR! I've updated the base from `master` to `develop`.

I'll take a look at this and test your changes.

**Comment by Sebi94nbg** [comment]
I've released version [3.0.22](https://github.com/barbushin/php-imap/releases/tag/3.0.22).

I've already tested everything with my test emails and everything looks fine. But there are a lot of different emails out, so I can not test / verify it to 100 %.

**Comment by Sebi94nbg** [comment]
Closing this since it seems to be fixed.
